core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala

```diff
@@ -0,0 +1,442 @@
+      } else {
+        // For ending timestamp the version should be before the provided timestamp.
+        val resolvedVersion = DeltaTableUtils.resolveTimeTravelVersion(
+          conf,
+          deltaLog,
+          spec
+        )
+        Some(resolvedVersion._1)
+      }
+    } else {
+      None
+    }
+  }
+
+  /**
+   * Get a Relation that represents change data between two snapshots of the table.
+   */
+  def getCDCRelation(
+      spark: SparkSession,
+      deltaLog: DeltaLog,
+      snapshotToUse: Snapshot,
+      partitionFilters: Seq[Expression],
+      conf: SQLConf,
+      options: CaseInsensitiveStringMap): BaseRelation = {
+
+    val startingVersion = getVersionForCDC(
+      spark,
+      deltaLog,
+      conf,
+      options,
+      DeltaDataSource.CDC_START_VERSION_KEY,
+      DeltaDataSource.CDC_START_TIMESTAMP_KEY
+    )
+
+    if (startingVersion.isEmpty) {
+      throw DeltaErrors.noStartVersionForCDC()
+    }
+
+    // add a version check here that is cheap instead of after trying to list a large version
+    // that doesn't exist
+    if (startingVersion.get > snapshotToUse.version) {
+      throw DeltaErrors.startVersionAfterLatestVersion(
+        startingVersion.get, snapshotToUse.version)
+    }
+
+    val endingVersion = getVersionForCDC(
+      spark,
+      deltaLog,
+      conf,
+      options,
+      DeltaDataSource.CDC_END_VERSION_KEY,
+      DeltaDataSource.CDC_END_TIMESTAMP_KEY
+    )
+
+    if (endingVersion.exists(_ < startingVersion.get)) {
+      throw DeltaErrors.endBeforeStartVersionInCDC(startingVersion.get, endingVersion.get)
+    }
+
+    logInfo(s"startingVersion: $startingVersion, endingVersion: $endingVersion")
+
+    DeltaCDFRelation(
+      cdcReadSchema(snapshotToUse.metadata.schema),
+      spark.sqlContext,
+      deltaLog,
+      startingVersion,
+      endingVersion
+    )
+  }
+
+  /**
+   * A special BaseRelation wrapper for CDF reads.
+   */
+  case class DeltaCDFRelation(
+      schema: StructType,
+      sqlContext: SQLContext,
+      deltaLog: DeltaLog,
+      startingVersion: Option[Long],
+      endingVersion: Option[Long]) extends BaseRelation with PrunedFilteredScan {
+
+    override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+      val df = changesToBatchDF(
+        deltaLog,
+        startingVersion.get,
+        endingVersion.getOrElse {
+          // If no ending version was specified, use the latest version as of scan building time.
+          // Note that this line won't be invoked (and thus we won't incur the update() cost)
+          // when endingVersion is present.
+          deltaLog.update().version
+        },
+        sqlContext.sparkSession)
+
+      df.select(requiredColumns.map(SchemaUtils.fieldNameToColumn): _*).rdd
+    }
+  }
+
+  /**
+   * For a sequence of changes(AddFile, RemoveFile, AddCDCFile) create a DataFrame that represents
+   * that captured change data between start and end inclusive.
+   *
```

Comment:
This is very long monolithic method. can you give sketch in this doc how it is building this DF? without that high level sketch its a little hard to make sense.

Reply:
actually.. alternatively. more inline comments explaining sections of the function will be better